File: bun-hono/src/modules/contact/contact.controller.ts
import AppError from "../../common/api.error";
import ContactService from "./contact.service";
import { Context } from "hono";
import ILogger from "../../config/logger/ilogger";
import { ApiResponse } from "../../common/api.response";
import { CreateContactSchema, CreateContactResponse, UpdateContactSchema, UpdateContactResponse, FindAllContactResponse } from "./contact.dto";

class ContactController {
  constructor(
    private contactService: ContactService,
    private logger: ILogger,
  ) {}

  async createContact(c: Context) {

    this.logger.debug("Controller Contact create Called")
    const {name, email, phone} = await c.req.json();
    const parseResult = CreateContactSchema.safeParse({name, email, phone});
    if (!parseResult.success) {
      throw new AppError(400, "Invalid request body");
    }
    const userId = c.get("userId");
    const result = await this.contactService.createContact(name, email, phone, userId);
    
    const response : ApiResponse<CreateContactResponse> = {
      message: "Contact created successfully",
      success: true,
      data: {
        id: result.id,
        userId: result.userId,
        name: result.name,
        email: result.email,
        phone: result.phone
      }
    }
    c.status(201)
    return c.json(response)
    
  }

  async findAllContact(c: Context) {
    
    const userId = c.get("userId")
    const contacts = await this.contactService.findAllContact(userId)
    const response : ApiResponse<FindAllContactResponse> = {
      message: "Contacts retrieved successfully",
      success: true,
      data: contacts
    }
    return c.json(response)

  }

  async updateContact(c: Context) {

    const userId = c.get("userId")
    const id = c.req.param('id')
    const {name, email, phone} = await c.req.json();
    const parseResult = UpdateContactSchema.safeParse({id, userId, name, email, phone});
  
    if (!parseResult.success){
      throw new AppError(400, "Invalid request body");
    }

    const result = await this.contactService.updateContact(id, userId, {name, email, phone});
    const response : ApiResponse<UpdateContactResponse> = {
      message: "Contact updated successfully",
      success: true,
      data: {
        id: result.id!,
        userId: result.userId!,
        name: result.name!,
        email: result.email!,
        phone: result.phone!
      }
    }
    c.status(200)
    return c.json(response);

  }

  async deleteContact(c: Context) {

    const id = c.req.param('id')
    const userId = c.get("userId")
    
    await this.contactService.deleteContact(id, userId)

    const resposne : ApiResponse<null> = {
        message: "Contact deleted successfully",
        success: true,
        data: null
      }
    c.status(200)
    return c.json(resposne)

  }
}

export default ContactController;

File: spring-java/src/main/java/org/distr/rcontacts/controller/ContactController.java
package org.distr.rcontacts.controller;

import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.RequestMapping;

import java.util.List;

import org.distr.rcontacts.dto.CreateContact;
import org.distr.rcontacts.dto.UpdateContactReq;
import org.distr.rcontacts.service.ContactService;
import org.distr.rcontacts.app.ApiResponse;
import org.distr.rcontacts.contracts.ContactContract;

@RestController
@RequestMapping("/api/v2")
public class ContactController {

    private ContactService contactService;

    public ContactController(ContactService contactService) {
        this.contactService = contactService;
    }

    @GetMapping("/contacts")
    public ResponseEntity<ApiResponse<List<ContactContract>>> getAllContacts() {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        String userId = (String) auth.getPrincipal();
        List<ContactContract> contacts = contactService.getAllContactsByUserId(userId);
        return ResponseEntity.ok(new ApiResponse<>("Contacts retrieved successfully", true, contacts));
    }

    @PostMapping("/contacts")
    public ResponseEntity<ApiResponse<ContactContract>>createContact(
        @RequestBody CreateContact createContactRequest
    ) {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        String userId = (String) auth.getPrincipal();
        ContactContract createdContact = contactService.CreateContact(
            userId,
            createContactRequest.name(),
            createContactRequest.phone(),
            createContactRequest.email()
        );
        return ResponseEntity.ok(new ApiResponse<>("Contact created successfully", true, createdContact));
    }

    @PutMapping("/contacts/{contactId}")
    public ResponseEntity<ApiResponse<ContactContract>>updateContact(
        @RequestBody UpdateContactReq updateContactRequest, 
        @PathVariable String contactId
    ) {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        String userId = (String) auth.getPrincipal();
        ContactContract updatedContact = contactService.updateContact(
            contactId,
            userId,
            updateContactRequest.name(),
            updateContactRequest.phone(),
            updateContactRequest.email()
        );
        return ResponseEntity.ok(new ApiResponse<>("Contact updated successfully", true, updatedContact));
    }

    @DeleteMapping("/contacts/{contactId}")
    public ResponseEntity<ApiResponse<Object>> deleteContact(@PathVariable String contactId) {
        contactService.deleteContact(contactId);
        return ResponseEntity.ok(new ApiResponse<>("Contact deleted successfully", true, null));
    }

}